### ISSUE TITLE
CSharpKernel: Possibility for a memory leak?

### ISSUE THREAD

**Issue by mikoskinen** [dotnet/interactive#2817]
### Describe the bug

We are using CSharpKernel through Microsoft.DotNet.Interactive.CSharp version 1.0.0-beta.22606.2 (also tested with 1.0.0-beta.23152.2) in our ASP.NET Core based application.

Lately there has been memory issues with the system and I've tried to pinpoint the origin. Through profiling I've noticed that things from Microsoft.CodeAnalysis namespace are using quite huge amounts of memory. I'm currently inclined to believe that there's maybe something going on in the latest versions of the CSharpKernel. 

Here's a simple .NET 7 Console app which has only this additional Nuget:

        <PackageReference Include="Microsoft.DotNet.Interactive.CSharp" Version="1.0.0-beta.22606.2" />

```
var kernel = new CSharpKernel();
var code = "var c = 15; c";

for (int i = 0; i < 1000; i++)
{
    await kernel.SendAsync(new SubmitCode(code), CancellationToken.None);
}
```

And here's how its memory usage behaves:

![image](https://user-images.githubusercontent.com/143570/224232683-5a957328-263a-4bf0-8f1c-8c06dacec200.png)

After 1000 runs, the app is using more than 5GB of memory. Now if we dig little deeper, this is what we see:

![image](https://user-images.githubusercontent.com/143570/224233032-c733bd55-1395-4e8c-8d23-19fcc8a9329a.png)

The Microsoft.CodeAnalysis.CommonReferenceManager seems to be the main problem. If we look at the call stack, here's what we can see:

![image](https://user-images.githubusercontent.com/143570/224236019-8766b0f4-8a14-4aa8-b9c5-a2433fc4207c.png)

I tried switching from kernel.SendAsync to kernel.SubmitCodeAsync but it is displaying the same:

![image](https://user-images.githubusercontent.com/143570/224234084-2994d8e6-feca-4eb5-8f1c-2b6313d52710.png)

The situation is better if the package Microsoft.DotNet.Interactive.CSharp is downgraded to 1.0.0-beta.22103.1. Now in the end the application takes little more than 1GB of memory:

![image](https://user-images.githubusercontent.com/143570/224235293-d266d1ad-560a-4870-8d64-f88bc94821eb.png)

The memory usage again goes back to more than 5GB if we upgrade to the latest version 1.0.0-beta.23152.2:

![image](https://user-images.githubusercontent.com/143570/224236511-2af899a3-857b-44e0-9dc5-7bd66e65de5d.png)

As a test I also tried a variant where new kernel is used for every execution and here's the result from that:

![image](https://user-images.githubusercontent.com/143570/224238257-09697ef6-18fc-4f76-96cd-093edb18a3f3.png)

**Comment by mikoskinen** [comment]
I cloned the interactive repo to test things out. It seems if I comment out this line:

https://github.com/dotnet/interactive/blob/66fe81b1c2711c39fdc703e9fd49819e9e671828/src/Microsoft.DotNet.Interactive.CSharp/InteractiveWorkspace.cs#L133

The memory usage drops to 1.0.0-beta.22103.1 levels, so from more 5GB to 1GB.

![image](https://user-images.githubusercontent.com/143570/224244027-e8a5830c-3c44-428c-b958-ee49d8d6f70a.png)

**Comment by jonsequitur** [comment]
Thank you for this detailed analysis, @mikoskinen! We're looking into it now.

**Comment by AbhitejJohn** [comment]
Fixed with #2866
